appl/include/obj_parser.h

```diff
@@ -1,181 +1,553 @@
+{
+    obj_info_t* obj_info = (obj_info_t*)malloc(sizeof(obj_info_t));
 
-            token = strtok_s(NULL, " ", &remaining_tokens);
-            obj->v[v_index].x = atof(token);
+    if(!obj_info)
+    {
+        return NULL;
+    }
 
-            token = strtok_s(NULL, " ", &remaining_tokens);
-            obj->v[v_index].y = atof(token);
+    obj_info->position_dynamic_array = NULL;
+    obj_info->position_amount = 0;
+    obj_info->position_dynamic_array_capacity = STARTING_DYNAMIC_ARRAY_CAPACITY;
+    obj_info->normal_dynamic_array = NULL;
+    obj_info->normal_amount = 0;
+    obj_info->normal_dynamic_array_capacity = STARTING_DYNAMIC_ARRAY_CAPACITY;
+    obj_info->uv_dynamic_array = NULL;
+    obj_info->uv_amount = 0;
+    obj_info->uv_dynamic_array_capacity = STARTING_DYNAMIC_ARRAY_CAPACITY;
+    obj_info->vertex_dynamic_array = NULL;
+    obj_info->vertex_amount = 0;
+    obj_info->vertex_dynamic_array_capacity = STARTING_DYNAMIC_ARRAY_CAPACITY;
+
+    return obj_info;
+}
 
-            token = strtok_s(NULL, " ", &remaining_tokens);
-            obj->v[v_index].z = atof(token);
+/// @brief Function to free the memory allocated for an obj_info
+/// @param obj_info The obj_info structure to free
+void __obj_info_free(obj_info_t* obj_info)
+{
+    free(obj_info->position_dynamic_array);
+    free(obj_info->normal_dynamic_array);
+    free(obj_info->uv_dynamic_array);
+    free(obj_info->vertex_dynamic_array);
 
-            v_index++;
-        }
+    free(obj_info);
+}
 
-        if (strncmp(buffer, "vt ", 3) == 0)
-        {
-            char *remaining_tokens;
-            char *token;
+/// @brief Function to create a new obj and initialize its members
+/// @return A pointer to the obj structure
+obj_t* __obj_new()
+{
+    obj_t* obj = (obj_t*)malloc(sizeof(obj_t));
+    if(!obj)
+    {
+        return NULL;
+    }
+
+    obj->triangles = NULL;
+    obj->triangle_amount = 0;
+    obj->triangles_capacity = STARTING_DYNAMIC_ARRAY_CAPACITY;
+    obj->vertex_amount = 0;
+    obj->position_amount = 0;
+    obj->uv_amount = 0;
+    obj->normal_amount = 0;
 
-            strtok_s(buffer, " ", &remaining_tokens); // vt token
+    return obj;
+}
 
-            token = strtok_s(NULL, " ", &remaining_tokens);
-            obj->vt[vt_index].x = atof(token);
+/// @brief Function to free the memory allocated for an obj
+/// @param obj The obj structure to free
+void __obj_free(obj_t* obj)
+{
+    free(obj->triangles);
 
-            token = strtok_s(NULL, " ", &remaining_tokens);
-            obj->vt[vt_index].y = atof(token);
+    free(obj);
+}
 
-            vt_index++;
-        }
+/// @brief Functiona to create a new triangle structure
+/// @param vertex1 The first vertex od the triangle
+/// @param vertex2 The second vertex od the triangle
+/// @param vertex3 The third vertex od the triangle
+/// @return A new triangle structure
+obj_triangle_t __obj_triangle_new(obj_vertex_t vertex1, obj_vertex_t vertex2, obj_vertex_t vertex3)
```

Comment:
this could be tricky in termes of data copying. The compiler could be smart enough to inline this function, but in the worse case this would happen:
- Each paramters is copied when function is invoked
- Then another copy of the parameters happen when assigning to v1, v2, v3
- Then a temporary result triangle is crated that will be copied in the array.

Looking at usage of this function (just one call), maybe could be better to inline the code where it is needed.

Reply:
Are you suggesting to do something like this? obj_triangle_t triangle; = (obj_triangle_t){vertex1, vertex2, vertex3};? Could you also please check discord message sent by me yesterday?

Reply:
nope. Basically remove that method and just use the internal code in the only one point where is needed